### PR TITLE
Added option to python_ta.check_all() to run black formatting tool

### PR DESCRIPTION
**Pull request by lana-w**
## Proposed Changes
Added an optional boolean parameter autoformat to python_ta.check_all so that, if True, causes the Python formatting tool black to run on each input file before it is checked with PythonTA.

## Type of Change

_(Write an `X` or a brief description next to the type or types that best describe your changes.)_

| Type                                                                                    | Applies? |
| --------------------------------------------------------------------------------------- | -------- |
| 🚨 _Breaking change_ (fix or feature that would cause existing functionality to change) |          |
| ✨ _New feature_ (non-breaking change that adds functionality)                          |X|
| 🐛 _Bug fix_ (non-breaking change that fixes an issue)                                  |          |
| ♻️ _Refactoring_ (internal change to codebase, without changing functionality)          |          |
| 🚦 _Test update_ (change that _only_ adds or modifies tests)                            |         |
| 📚 _Documentation update_ (change that _only_ updates documentation)                    |          |
| 📦 _Dependency update_ (change that updates a dependency)                               |X|
| 🔧 _Internal_ (change that _only_ affects developers or continuous integration)         |          |

## Checklist

_(Complete each of the following items for your pull request. Indicate that you have completed an item by changing the `[ ]` into a `[x]` in the raw text, or by clicking on the checkbox in the rendered description on GitHub.)_

Before opening your pull request:

- [X] I have performed a self-review of my changes.
  - Check that all changed files included in this pull request are intentional changes.
  - Check that all changes are relevant to the purpose of this pull request, as described above.
- [X] I have added tests for my changes, if applicable.
  - This is **required** for all bug fixes and new features.
- [X] I have updated the project documentation, if applicable.
  - This is **required** for new features.
- [x] I have updated the project Changelog (this is required for all changes).
- [ ] If this is my first contribution, I have added myself to the list of contributors.

After opening your pull request:

- [x] I have verified that the pre-commit.ci checks have passed.
- [x] I have verified that the CI tests have passed.
- [ ] I have reviewed the test coverage changes reported by Coveralls.
- [x] I have [requested a review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) from a project maintainer.

## Questions and Comments
Leaving this as a draft while I work on some tests for this feature, as have only tested on individual files so far rather than folders. Regarding adding dependencies, is there a specific version of black that is required? It seems they promise stability for each calendar year: https://black.readthedocs.io/en/stable/the_black_code_style/index.html#stability-policy

Have also not updated any documentation yet, but wanted to give an update on the functionality's progress!

June 13 Update: I have made and locally tested the changes you asked for regarding how black is called!

I have written a sample test for the autoformat parameter, and am looking for feedback in terms of efficiency before I add a couple more to check that line length works correctly. Currently, it makes a temporary file, runs python_ta.check_all on it, and then removes it, which is fairly slow. 

I have also made a update to the configuration.md documentation as a best guess, so let me know if there's any other places to add new documentation.

Currently trying to figure out why the PR tests are failing!

### CODE REVIEW

**Comment by david-yz-liu** [comment]
@lana-w great work so far. Don't worry about specifying a particular Black version. Instead, our documentation should say that the `autoformat` feature is compatible with whatever version of `black` is installed.

A few other comments:

- use `sys.executable` to specify the python version, then `-m black` to actually run black, rather than just `black` itself
- PythonTA can specify a max line length for its checks; that same max line length should be passed as a command-line argument to black for its formatting purposes
- also use the `--skip-string-normalization` command-line argument for black

**Comment by lana-w** [comment]
@david-yz-liu Have made these updates, and left an update on status in the comments part of the PR!

**Comment by david-yz-liu** [comment]
@lana-w thanks for the update, and yes the testing approach you've used is very appropriate. You should also capture the reporter output and check that there are no `pep8-errors` message.

You can also add an additional test case that uses the same temporary file approach, but with `autoformat=False`, and check that the file's contents are unchanged.

File: tests/test_black.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        f.write(source)
+    return file_path
+
+
+@pytest.mark.parametrize("source, pep8_check, expected_code", error_params)
+def test_black_formatting(source, pep8_check, expected_code, unformatted_file):
+    output = io.StringIO()
+
+    with contextlib.redirect_stdout(output):
+        result = subprocess.run(
```
Comment (david-yz-liu) [inline tests/test_black.py:46]: Overall, you shouldn't need to use `subprocess` here, and removing it may make it easier to debug the timeout issue. Call `python_ta.check_all` direction in here, and pass in the path to the temporary file.

File: python_ta/__init__.py [hunk 2]
```diff
@@ -191,9 +213,9 @@ def _check(
                 )
             if linter.config.pyta_error_permission:
                 errs = list(current_reporter.messages.values())
-            if (
-                f_paths != [] or errs != []
-            ):  # Only call upload_to_server() if there's something to upload
+
+            # Only call upload_to_server() if there's something to upload
```
Comment (david-yz-liu) [inline python_ta/__init__.py:217]: Revert this change

File: python_ta/__init__.py [hunk 3]
```diff
@@ -204,6 +226,7 @@ def _check(
                     url=linter.config.pyta_server_address,
                     version=__version__,
                 )
+
```
Comment (david-yz-liu) [inline python_ta/__init__.py:229]: Revert this change

File: python_ta/__init__.py [hunk 4]
```diff
@@ -83,6 +84,7 @@ def check_all(
     config: Union[dict, str] = "",
     output: Optional[TextIO] = None,
     load_default_config: bool = True,
+    autoformat: Optional[bool] = False,
```
Comment (david-yz-liu) [inline python_ta/__init__.py:87]: Also add to `check_errors`

File: CHANGELOG.md [hunk 5]
```diff
@@ -61,6 +62,7 @@ list only contains the Pylint checkers enabled by default in PythonTA.
 - Updated dependencies for GitHub Actions to use the latest versions
 - Updated dependabot configuration to auto-update dependencies for GitHub Actions in the future
 - Updated usage messages in `examples/sample_usage/` of `draw_cfg.py` and `print_ast.py` to be accurate on all operating systems
+- Added option that runs black formatting tool to python_ta.check_all
```
Comment (david-yz-liu) [inline CHANGELOG.md:65]: This isn't an internal change (duplicated entry from above)

**Comment by lana-w** [comment]
Have made all the requested updates, and removed the usage of subprocess.run in the tests!

File: python_ta.egg-info/requires.txt [hunk 6]
```diff
@@ -0,0 +1,34 @@
+astroid~=3.1.0
```
Comment (david-yz-liu) [inline python_ta.egg-info/requires.txt:1]: Remove this file (it should not be committed, and probably you can delete it from your local repository as well)

File: python_ta/__init__.py [hunk 7]
```diff
@@ -191,6 +217,7 @@ def _check(
                 )
             if linter.config.pyta_error_permission:
                 errs = list(current_reporter.messages.values())
+
```
Comment (david-yz-liu) [inline python_ta/__init__.py:220]: Revert this change

File: CHANGELOG.md [hunk 8]
```diff
@@ -20,6 +20,7 @@ and adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 - Stored actual AST condition node in edges leading out of If/While blocks in generated control flow graphs.
 - Stored valid Python function preconditions in initial edge to function code in generated function control flow graphs.
 - Report warning when control flow graph creation encounters a syntax error related to control flow
+- Added option that runs black formatting tool to python_ta.check_all()
```
Comment (david-yz-liu) [inline CHANGELOG.md:23]: Provide the name of the option here ("Added `autoformat` option for running the black formatting tool...")

**Comment by lana-w** [comment]
@david-yz-liu Resolved!

**Review by david-yz-liu — APPROVED** [review]
